src/Microsoft.Data/PrimitiveDataFrameColumnContainer.cs

```diff
@@ -0,0 +1,158 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Data
+{
+    public partial class PrimitiveDataFrameColumnContainer<T>
+        where T : struct
+    {
+        public IList<DataFrameBuffer<T>> Buffers = new List<DataFrameBuffer<T>>();
+        public PrimitiveDataFrameColumnContainer(T[] values)
+        {
+            values = values ?? throw new ArgumentNullException(nameof(values));
+            long ii = 0;
+            long length = values.LongLength;
+            DataFrameBuffer<T> curBuffer;
+            if (Buffers.Count == 0)
+            {
+                curBuffer = new DataFrameBuffer<T>();
+                Buffers.Add(curBuffer);
+            }
+            else
+            {
+                curBuffer = Buffers[Buffers.Count - 1];
+            }
+            for (; ii < length; ii++)
+            {
+                if (curBuffer.Length == curBuffer.MaxCapacity)
+                {
+                    curBuffer = new DataFrameBuffer<T>();
+                    Buffers.Add(curBuffer);
+                }
+                curBuffer.Append(values[ii]);
+                Length++;
+            }
+        }
+        public PrimitiveDataFrameColumnContainer(IEnumerable<T> values)
+        {
+            values = values ?? throw new ArgumentNullException(nameof(values));
+            if (Buffers.Count == 0)
+            {
+                Buffers.Add(new DataFrameBuffer<T>());
+            }
+            var curBuffer = Buffers[Buffers.Count - 1];
+            foreach (var value in values)
+            {
+                if (curBuffer.Length == curBuffer.MaxCapacity)
+                {
+                    curBuffer = new DataFrameBuffer<T>();
+                    Buffers.Add(curBuffer);
+                }
+                curBuffer.Append(value);
+                Length++;
+            }
+        }
+        public PrimitiveDataFrameColumnContainer() { }
+        public long Length;
+        //TODO:
+        public long NullCount => throw new NotImplementedException();
+        private int GetArrayContainingRowIndex(ref long rowIndex)
+        {
+            if (rowIndex > Length)
+            {
+                throw new ArgumentException($"Index {rowIndex} cannot be greater than the Column's Length {Length}");
+            }
+            int curArrayIndex = 0;
+            int numBuffers = Buffers.Count;
+            while (curArrayIndex < numBuffers && rowIndex > Buffers[curArrayIndex].Length)
+            {
+                rowIndex -= Buffers[curArrayIndex].Length;
+                curArrayIndex++;
+            }
+            return curArrayIndex;
+        }
+
+        public IList<T> this[long startIndex, int length]
+        {
+            get
+            {
+                var ret = new List<T>();
```

Comment:
You should initialize this list to the length capacity. That way it doesn't have to grow while you are adding elements to it.